extensions/oidc-keycloak/deployment/src/main/java/io/quarkus/devservices/keycloak/deployment/KeycloakDevServicesProcessor.java

```diff
@@ -0,0 +1,410 @@
+import io.quarkus.deployment.builditem.LaunchModeBuildItem;
+import io.quarkus.deployment.builditem.RunTimeConfigurationDefaultBuildItem;
+import io.quarkus.deployment.builditem.ServiceStartBuildItem;
+import io.quarkus.runtime.LaunchMode;
+import io.quarkus.runtime.configuration.ConfigUtils;
+import io.vertx.core.Vertx;
+import io.vertx.core.http.HttpHeaders;
+import io.vertx.mutiny.core.buffer.Buffer;
+import io.vertx.mutiny.ext.web.client.WebClient;
+
+public class KeycloakDevServicesProcessor {
+    public static volatile DevServicesConfig capturedDevServicesConfiguration;
+    static volatile Vertx vertxInstance;
+
+    private static final Logger LOG = Logger.getLogger(KeycloakDevServicesProcessor.class);
+
+    private static final String CONFIG_PREFIX = "quarkus.oidc.";
+    private static final String AUTH_SERVER_URL_CONFIG_KEY = CONFIG_PREFIX + "auth-server-url";
+    private static final String APPLICATION_TYPE_CONFIG_KEY = CONFIG_PREFIX + "application-type";
+    private static final String CLIENT_ID_CONFIG_KEY = CONFIG_PREFIX + "client-id";
+    private static final String CLIENT_SECRET_CONFIG_KEY = CONFIG_PREFIX + "credentials.secret";
+    private static final String KEYCLOAK_URL_KEY = "keycloak.url";
+
+    private static final int KEYCLOAK_EXPOSED_PORT = 8080;
+    private static final String KEYCLOAK_DOCKER_REALM_PATH = "/tmp/realm.json";
+    private static final String KEYCLOAK_USER_PROP = "KEYCLOAK_USER";
+    private static final String KEYCLOAK_PASSWORD_PROP = "KEYCLOAK_PASSWORD";
+    private static final String KEYCLOAK_VENDOR_PROP = "DB_VENDOR";
+    private static final String KEYCLOAK_IMPORT_PROP = "KEYCLOAK_IMPORT";
+    private static final String KEYCLOAK_ADMIN_USER = "admin";
+    private static final String KEYCLOAK_ADMIN_PASSWORD = "admin";
+    private static final String KEYCLOAK_DB_VENDOR = "H2";
+    private static final String OIDC_USERS = "oidc.users";
+
+    private static volatile List<Closeable> closeables;
+    private static volatile boolean first = true;
+    private final IsDockerWorking isDockerWorking = new IsDockerWorking(true);
+
+    KeycloakBuildTimeConfig config;
+
+    @BuildStep(onlyIfNot = IsNormal.class)
+    public DevServicesConfigBuildItem startKeycloakContainer(LaunchModeBuildItem launchMode,
+            BuildProducer<RunTimeConfigurationDefaultBuildItem> runTimeConfiguration,
+            BuildProducer<ServiceStartBuildItem> serviceStartBuildItemBuildProducer,
+            BuildProducer<DevServicesNativeConfigResultBuildItem> devServices) {
+
+        DevServicesConfig currentDevServicesConfiguration = config.devservices;
+        // Figure out if we need to shut down and restart any existing Keycloak container
+        // if not and the Keycloak container has already started we just return
+        if (closeables != null) {
+            boolean restartRequired = launchMode.getLaunchMode() == LaunchMode.DEVELOPMENT
```

Comment:
This should not factor in the launch mode, it should just be based on if the config has changed (but the 'config changed' check should be expanded to include if the json import file changes). We only want to restart the container if it is absolutely necessary, because it is slow.

Note that this behaviour does vary a bit between the DevServices. Mongo and Redis in particular are restarted to make sure there is no left over data. The SQL databases are not as hibernate knows how to clean out all the data between test runs. At some point this restart behavior for the other ones may be improved, because in general you want to restart as little as possible.

Reply:
Hi @stuartwdouglas OK, I'll remove the `DEVELOPMENT` check, makes sense; but as I commented [here](https://github.com/quarkusio/quarkus/pull/17364#discussion_r668014434), I believe we don't need to check the realm file changes - since those changes can only happen if the user (re-)exports whatever has been updated in the live Keycloak instance's console - but these changes are indirectly picked by the Dev UI which requests the tokens from KC.